xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-macros/xwiki-platform-rendering-macro-code-oldcore/src/main/java/org/xwiki/rendering/internal/macro/code/source/DocumentObjectPropertyCodeMacroSourceLoader.java

```diff
@@ -0,0 +1,115 @@
+import com.xpn.xwiki.objects.PropertyInterface;
+import com.xpn.xwiki.objects.classes.BaseClass;
+import com.xpn.xwiki.objects.classes.EmailClass;
+import com.xpn.xwiki.objects.classes.PasswordClass;
+import com.xpn.xwiki.objects.classes.TextAreaClass;
+import com.xpn.xwiki.objects.classes.TextAreaClass.ContentType;
+
+/**
+ * @version $Id$
+ * @since 15.0RC1
+ * @since 14.10.2
+ */
+@Component(hints = {"OBJECT_PROPERTY", "PAGE_OBJECT_PROPERTY"})
+@Singleton
+public class DocumentObjectPropertyCodeMacroSourceLoader implements EntityCodeMacroSourceLoader
+{
+    @Inject
+    private GeneralMailConfiguration mailConfiguration;
+
+    @Override
+    public CodeMacroSource load(XWikiDocument document, EntityReference entityReference,
+        CodeMacroSourceReference reference, XWikiContext xcontext) throws MacroExecutionException
+    {
+        BaseObject xobject = document.getXObject(entityReference.getParent());
+        BaseProperty xobjectProperty;
+        try {
+            xobjectProperty = (BaseProperty) xobject.get(entityReference.getName());
+        } catch (XWikiException e) {
+            throw new MacroExecutionException("Failed to access object property [" + entityReference + "]", e);
+        }
+
+        if (xobjectProperty == null) {
+            throw new MacroExecutionException("Unknown property [" + entityReference + "]");
+        }
+
+        return new CodeMacroSource(reference, xobjectProperty.toText(),
+            getLanguage(xobject, document, entityReference, xcontext));
+    }
+
+    private String getLanguage(BaseObject xobject, XWikiDocument document, EntityReference entityReference,
+        XWikiContext xcontext) throws MacroExecutionException
+    {
+        BaseClass xclass = xobject.getXClass(xcontext);
+
+        if (xclass != null) {
+            PropertyInterface xclassProperty = xclass.get(entityReference.getName());
+
+            // Displaying a password is forbidden
+            if (xclassProperty instanceof PasswordClass) {
+                throw new MacroExecutionException(String.format(
+                    "Displaying content of property [%s] is not allowed because it's a passwordl", entityReference));
+            }
+
+            // Displaying email is forbidden when obfuscation is enabled
+            if (xclassProperty instanceof EmailClass && this.mailConfiguration.shouldObfuscate()) {
+                throw new MacroExecutionException(
+                    String.format("Displaying content of property [%s] is not allowed because it's an obfuscated email",
+                        entityReference));
+            }
+
+            String language = null;
+            if (xclassProperty instanceof TextAreaClass) {
+                TextAreaClass textarea = (TextAreaClass) xclassProperty;
+
+                ContentType contentType = ContentType.getByValue(textarea.getContentType());
+                if (contentType == ContentType.VELOCITY_CODE || contentType == ContentType.VELOCITYWIKI) {
```

Comment:
I'm not sure to understand why we need to handle velocity with a special case here.

Reply:
Well, I am not sure to understand what you don't understand actually :)

The point of this code is to find the language (from code macro point of view) of the content, and when the content type is "VELOCITY_CODE" and "VELOCITYWIKI" we know it's velocity so we check it.

Reply:
ok, got it on a second read. Thanks.